hermes-client/src/main/java/pl/allegro/tech/hermes/client/ReactiveHermesClient.java

```diff
@@ -0,0 +1,314 @@
+
+    ReactiveHermesClient(ReactiveHermesSender sender,
+                         URI uri,
+                         Map<String, String> defaultHeaders,
+                         int maxRetries,
+                         Predicate<HermesResponse> retryCondition,
+                         long retrySleepInMillis,
+                         long maxRetrySleepInMillis,
+                         ScheduledExecutorService scheduler) {
+        this.sender = sender;
+        this.uri = createUri(uri);
+        this.defaultHeaders = Collections.unmodifiableMap(new HashMap<>(defaultHeaders));
+        this.maxRetries = maxRetries;
+        this.retryCondition = retryCondition;
+        this.retrySleep = Duration.ofMillis(retrySleepInMillis);
+        this.maxRetrySleep = Duration.ofMillis(maxRetrySleepInMillis);
+        this.scheduler = Schedulers.fromExecutor(scheduler);
+    }
+
+    private String createUri(URI uri) {
+        String uriString = uri.toString();
+        return uriString + (uriString.endsWith("/") ? "" : "/") + "topics/";
+    }
+
+    public Mono<HermesResponse> publishJSON(String topic, byte[] message) {
+        return publish(hermesMessage(topic, message).json().build());
+    }
+
+    public Mono<HermesResponse> publishJSON(String topic, String message) {
+        return publish(hermesMessage(topic, message).json().build());
+    }
+
+    public Mono<HermesResponse> publishAvro(String topic, int schemaVersion, byte[] message) {
+        return publish(hermesMessage(topic, message).avro(schemaVersion).build());
+    }
+
+    public Mono<HermesResponse> publish(String topic, String message) {
+        return publish(hermesMessage(topic, message).build());
+    }
+
+    public Mono<HermesResponse> publish(String topic, String contentType, byte[] message) {
+        return publish(hermesMessage(topic, message).withContentType(contentType).build());
+    }
+
+    public Mono<HermesResponse> publish(String topic, String contentType, String message) {
+        return publish(hermesMessage(topic, message).withContentType(contentType).build());
+    }
+
+    public Mono<HermesResponse> publish(String topic, String contentType, int schemaVersion, byte[] message) {
+        return publish(hermesMessage(topic, message).withContentType(contentType).withSchemaVersion(schemaVersion).build());
+    }
+
+    public Mono<HermesResponse> publish(HermesMessage message) {
+        if (shutdown) {
+            return completedWithShutdownException();
+        }
+        HermesMessage.appendDefaults(message, defaultHeaders);
+        return publishWithRetries(message);
+    }
+
+    public boolean addMessageDeliveryListener(MessageDeliveryListener listener) {
+        return messageDeliveryListeners.add(listener);
+    }
+
+    private Mono<HermesResponse> publishWithRetries(HermesMessage message) {
+        currentlySending.incrementAndGet();
+
+        Retry retry = prepareRetry(message);
+
+        return sendOnce(message)
```

Comment:
this chain is huuge ;) can we divide it to the smaller parts ? I can see here few of them:
1. getting attempt result based on response
2. mapping results to Mono (with result or with exception in case of failed result)
3. handling exhausted retries